src/qibo/result.py

```diff
@@ -237,6 +247,36 @@ def frequencies(self, binary=True, registers=False):
 
         return self._frequencies
 
+    def probabilities(self, qubits=None):
+        """Calculate the probabilities as frequencies / nshots
+
+        Returns:
+            The array containing the probabilities of the measured qubits.
+        """
+        nqubits = len(self.measurement_gate.qubits)
+        if qubits is None:
+            qubits = range(nqubits)
+        else:
+            if not set(qubits).issubset(self.measurement_gate.qubits):
+                raise_error(
+                    RuntimeError,
+                    f"Asking probabilities for qubits {qubits}, but only qubits {self.measurement_gate.qubits} were measured.",
+                )
+
+        if self._probs is not None:
+            return self.backend.calculate_probabilities(
+                self._probs, qubits, len(self.measurement_gate.qubits)
+            )
+
+        frequencies = self.frequencies(binary=False)
+        probs = self.backend.calculate_probabilities_from_frequencies(
+            frequencies, self._samples.shape[1], self.nshots
```

Comment:
> Would
if qubits is None:
    qubits = self.measurement_gate.qubits
solve this issue? I understand that probabilities from frequencies only exists for MeasurementOutcomes where you always have some measured qubits. This set of measured qubits is the maximum you can calculate probabilities for so I think this choice makes sense.

The problem is that you don't have access to the measurements, and thus to which qubits are measured, in `QuantumState` though. Therefore, if you don't specify which qubits to trace out to the `QuantumState.probabilities()` you will get the probabilities over all the qubits, whereas for `MeasurementOutcomes.probabilities()` you get the probabilities over the measured qubits only. I am not sure this is a problem, but it is an asymmetry.

> I have not tried this, but I suspect this will fail in the case you only have frequencies and not samples. For example if the user asks for frequencies and then for probabilities, without asking for samples in between or before.

If I am not mistaken, the calculation of the frequencies always involves the generation of the samples first, unless you have the probabilities available because you constructed the `MeasurementOutcomes` object starting from those. At that point, though, you will not call the `backend.calculate_probabilities_from_frequencies`.